File: backend/models/product.go
// ------------------
// models/product.go
// ------------------
package models

import (
	"gorm.io/datatypes"
	"gorm.io/gorm"
)

type Attribute struct {
	Name  string `json:"name"`
	Value string `json:"value"`
}

type Product struct {
	gorm.Model
	Tags        datatypes.JSON `json:"tags" gorm:"type:json"`
	Attributes  datatypes.JSON `json:"attributes" gorm:"type:json"`
	Name        string         `json:"name"`
	Description string         `json:"description"`
	Price       float64        `json:"price"`
	ImageURL    string         `json:"imageUrl"`
	InMainPage  bool           `json:"inMainPage"`
	CategoryID  uint           `json:"categoryID"`
	Category    Category
	Reviews     []Review `json:"review" gorm:"foreignKey:ProductID"`
}

File: front-end/app/types/product.ts
// types/product.ts
// تعریف نوع Attribute
export interface Attribute {
  name: string;
  value: string;
}

// تعریف نوع Category
export interface Category {
  ID: number;
  CreatedAt: string;
  UpdatedAt: string;
  DeletedAt: string | null;
  name: string;
  Products: Product[] | null;
}

// تعریف نوع اصلی Product
export interface Product {
  ID: number;
  CreatedAt: string;
  UpdatedAt: string;
  DeletedAt: string | null;
  tags: string[];
  attributes: Attribute[];
  name: string;
  description: string;
  price: number;
  imageUrl: string;
  inMainPage: boolean;
  categoryID: number;
  Category: Category | null;
  review: Review[];
}

// یا به صورت کامل‌تر با استفاده از Generic و Partial برای عملیات CRUD
export interface BaseEntity {
  ID: number;
  CreatedAt: string;
  UpdatedAt: string;
  DeletedAt: string | null;
}

export interface ProductEntity extends BaseEntity {
  tags: string[];
  attributes: Attribute[];
  name: string;
  description: string;
  price: number;
  imageUrl: string;
  inMainPage: boolean;
  categoryID: number;
  Category: CategoryEntity | null;
}

export interface CategoryEntity extends BaseEntity {
  name: string;
  Products: ProductEntity[] | null;
}

export interface Review {
  id: number;
  name: string;
  email: string;
  comment: string;
}
